src/ParameterBag.php

```diff
@@ -0,0 +1,95 @@
+<?php declare(strict_types=1);
+
+namespace OpenSourceRefinery\Component\QueryLanguage;
+
+use OpenSourceRefinery\Component\QueryLanguage\Types\IntegerValue;
+use OpenSourceRefinery\Component\QueryLanguage\Types\InvalidParameterValue;
+use OpenSourceRefinery\Component\QueryLanguage\Types\NotSupportedParameter;
+use OpenSourceRefinery\Component\QueryLanguage\Types\ParameterValue;
+use OpenSourceRefinery\Component\QueryLanguage\Types\StringValue;
+
+final class ParameterBag
+{
+    /**
+     * @var ParameterValue[] Mapped by id in format x.y.z for easy access
+     */
+    private $values;
+
+    /**
+     * @param ParameterValue[] Indexed by string key
+     */
+    private function __construct(array $values)
+    {
+        $this->values = $values;
+    }
+
+    public function has(string $parameter): bool
+    {
+        if (! \array_key_exists($parameter, $this->values)) {
+            // search by key parts, starting with root node (root.key.value)
+            $keys = \array_keys($this->values);
+            return \count(
+                \array_filter(
+                    $keys,
+                    function (string $key) use ($parameter) {
+                        return \strpos($key, $parameter) === 0;
+                    }
+                )
+            ) > 0;
+        }
+
+        return true;
+    }
+
+    public function get(string $parameter): ParameterValue
+    {
+        if (! $this->has($parameter)) {
+            throw new NotSupportedParameter($parameter);
+        }
+
+        return $this->values[$parameter];
+    }
+
+    public static function fromArray(array $array): self
+    {
+        $closure = function (array $_flattened, string $current_node, $value, callable $closure): array {
+            if (\is_array($value) && \count($value) > 0) {
+                foreach ($value as $k => $v) {
+                    $_flattened = $closure($_flattened, $current_node . '.' . $k, $v, $closure);
```

Comment:
can you explain why we are flattening?

Reply:
I wanted to be able to use the ParameterBag in the simplest way. 
ie.

`$bag->has('some.key')`, instead of foreaching on the keys and when the key `some` is found then iterate on other keys to find `key`. I found that is was easier for the parser to use the bag that can find somthing fast, without too much iteration